File: python/src/linksql/store.py
"""In-memory doublets store.

The associative data model used by LinksQL is built from a single primitive: the
**link** (a doublet). Every link is a triple::

    (index: source target)

where ``index`` is the link's own unique identity and ``source``/``target`` are
the indices of the linked links. A link whose source and target both equal its
own index is a *point* — the atomic node of the graph.

The store owns identity allocation and enforces the associative invariant that a
``(source, target)`` pair identifies at most one link (deduplication). It is a
pure data container: pattern matching and the substitution operation live in
:mod:`linksql.substitution`, which only relies on the read/write surface defined
here.

This module is a faithful, behaviourally-identical port of ``src/store.js`` from
the JavaScript reference implementation.
"""

from __future__ import annotations

from dataclasses import dataclass


class LinkIntegrityError(Exception):
    """Error raised when a store operation would violate model integrity."""


@dataclass
class Link:
    """A single link: a triple ``(index: source target)`` of positive integers."""

    index: int
    source: int
    target: int


def _is_integer(value: object) -> bool:
    """Return whether ``value`` is an integer (and not a ``bool``).

    JavaScript's ``Number.isInteger`` rejects non-numbers; in Python ``bool`` is
    a subclass of ``int``, so exclude it explicitly to keep the integrity checks
    faithful (a boolean is never a valid index).
    """
    return isinstance(value, int) and not isinstance(value, bool)


class LinksStore:
    """A mutable set of links addressed by integer identity."""

    def __init__(self) -> None:
        # Insertion order is preserved by Python's ``dict`` natively, matching
        # the JavaScript ``Map`` iteration order used by ``all()``.
        self.links: dict[int, Link] = {}
        self.by_pair: dict[tuple[int, int], int] = {}
        # Next identity to hand out; kept above every used index.
        self.next_index: int = 1

    @property
    def size(self) -> int:
        """Number of stored links."""
        return len(self.links)

    def has(self, index: int) -> bool:
        """Return whether a link with this identity exists."""
        return index in self.links

    def get(self, index: int) -> Link | None:
        """Fetch a link by identity, or ``None`` when absent."""
        return self.links.get(index)

    def find_by_pair(self, source: int, target: int) -> Link | None:
        """Look up a link by its ``(source, target)`` structure."""
        index = self.by_pair.get((source, target))
        if index is None:
            return None
        return self.links.get(index)

    def all(self) -> list[Link]:
        """Return all links in insertion order."""
        return list(self.links.values())

    def allocate_index(self) -> int:
        """Reserve a fresh identity, advancing the allocator past it."""
        while self.next_index in self.links:
            self.next_index += 1
        index = self.next_index
        self.next_index += 1
        return index

    def reserve_index(self, index: int) -> None:
        """Keep the allocator above an externally chosen identity."""
        if index >= self.next_index:
            self.next_index = index + 1

    def create(
        self,
        *,
        source: int,
        target: int,
        index: int | None = None,
    ) -> Link:
        """Create (or, by deduplication, return) a link."""
        if not _is_integer(source) or not _is_integer(target):
            message = "Link source and target must be integers"
            raise LinkIntegrityError(message)
        existing = self.find_by_pair(source, target)
        if existing is not None:
            if index is not None and index != existing.index:
                message = (
                    f"Link ({source} {target}) already exists as "
                    f"{existing.index}, cannot also be {index}"
                )
                raise LinkIntegrityError(message)
            return existing
        if index is None:
            link_id = self.allocate_index()
        else:
            if not _is_integer(index) or index < 1:
                message = "Link index must be a positive integer"
                raise LinkIntegrityError(message)
            if index in self.links:
                message = f"Link index {index} is already in use"
                raise LinkIntegrityError(message)
            self.reserve_index(index)
            link_id = index
        link = Link(index=link_id, source=source, target=target)
        self.links[link_id] = link
        self.by_pair[(source, target)] = link_id
        return link

    def update(
        self,
        index: int,
        *,
        source: int,
        target: int,
        new_index: int | None = None,
    ) -> Link:
        """Replace the structure of an existing link, preserving its identity.

        The identity is preserved unless ``new_index`` requests a re-index.
        """
        current = self.links.get(index)
        if current is None:
            message = f"Cannot update missing link {index}"
            raise LinkIntegrityError(message)
        link_id = index if new_index is None else new_index
        collision = self.find_by_pair(source, target)
        if collision is not None and collision.index != index:
            message = (
                f"Cannot update link {index}: ({source} {target}) already "
                f"exists as {collision.index}"
            )
            raise LinkIntegrityError(message)
        if link_id != index and link_id in self.links:
            message = f"Link index {link_id} is already in use"
            raise LinkIntegrityError(message)
        del self.by_pair[(current.source, current.target)]
        del self.links[index]
        link = Link(index=link_id, source=source, target=target)
        self.links[link_id] = link
        self.by_pair[(source, target)] = link_id
        self.reserve_index(link_id)
        return link

    def delete(self, index: int) -> bool:
        """Remove a link by identity; return whether a link was removed."""
        current = self.links.get(index)
        if current is None:
            return False
        del self.by_pair[(current.source, current.target)]
        del self.links[index]
        return True

    def clear(self) -> None:
        """Remove every link and reset identity allocation."""
        self.links.clear()
        self.by_pair.clear()
        self.next_index = 1

File: csharp/src/LinksQL/Store.cs
using System.Globalization;

namespace LinksQL;

/// <summary>
/// A link (doublet): a triple of integer identities.
/// </summary>
/// <param name="Index">The link's own unique identity.</param>
/// <param name="Source">The index this link points from.</param>
/// <param name="Target">The index this link points to.</param>
/// <remarks>
/// A link whose <paramref name="Source"/> and <paramref name="Target"/> both
/// equal its own <paramref name="Index"/> is a <em>point</em> — the atom of the
/// model.
/// </remarks>
public sealed record Link(long Index, long Source, long Target);

/// <summary>
/// In-memory doublets store: a mutable set of links addressed by integer
/// identity.
/// </summary>
/// <remarks>
/// <para>
/// The store owns identity allocation and enforces the associative invariant
/// that a <c>(source, target)</c> pair identifies at most one link
/// (deduplication). It is a pure data container: pattern matching and the
/// substitution operation live in <see cref="Substitution"/>.
/// </para>
/// <para>
/// <see cref="All"/> preserves insertion order (matching the JavaScript
/// <c>Map</c> iteration order) via a parallel order list; links are
/// <em>not</em> sorted by index. As in JavaScript, <see cref="Update"/> removes
/// and re-inserts the entry, so an updated link always moves to the end of
/// iteration order.
/// </para>
/// </remarks>
public sealed class LinksStore
{
    private readonly Dictionary<long, Link> _links = new();
    private readonly List<long> _order = new();
    private readonly Dictionary<(long Source, long Target), long> _byPair = new();

    /// <summary>Gets the next identity to hand out; kept above every used index.</summary>
    public long NextIndex { get; private set; } = 1;

    /// <summary>Gets the number of stored links.</summary>
    public int Size => _links.Count;

    /// <summary>Gets the number of stored links.</summary>
    public int Count => _links.Count;

    /// <summary>Whether a link with this identity exists.</summary>
    /// <param name="index">Identity to test.</param>
    /// <returns><see langword="true"/> if a link with this identity exists.</returns>
    public bool Has(long index) => _links.ContainsKey(index);

    /// <summary>Fetch a link by identity.</summary>
    /// <param name="index">Identity to fetch.</param>
    /// <returns>The link, or <see langword="null"/> when absent.</returns>
    public Link? Get(long index) => _links.TryGetValue(index, out var link) ? link : null;

    /// <summary>Look up a link by its <c>(source, target)</c> structure.</summary>
    /// <param name="source">Source index.</param>
    /// <param name="target">Target index.</param>
    /// <returns>The link, or <see langword="null"/> when absent.</returns>
    public Link? FindByPair(long source, long target) =>
        _byPair.TryGetValue((source, target), out var index) ? _links[index] : null;

    /// <summary>All links, in insertion order.</summary>
    /// <returns>An ordered list of every stored link.</returns>
    public IReadOnlyList<Link> All()
    {
        var result = new List<Link>(_order.Count);
        foreach (var index in _order)
        {
            result.Add(_links[index]);
        }

        return result;
    }

    /// <summary>Reserve a fresh identity, advancing the allocator past it.</summary>
    /// <returns>A previously-unused index.</returns>
    public long AllocateIndex()
    {
        while (_links.ContainsKey(NextIndex))
        {
            NextIndex += 1;
        }

        return NextIndex++;
    }

    /// <summary>Keep the allocator above an externally chosen identity.</summary>
    /// <param name="index">An index that is now in use.</param>
    public void ReserveIndex(long index)
    {
        if (index >= NextIndex)
        {
            NextIndex = index + 1;
        }
    }

    /// <summary>
    /// Create (or, by deduplication, return) a link.
    /// </summary>
    /// <param name="index">
    /// Explicit identity; auto-allocated when <see langword="null"/>.
    /// </param>
    /// <param name="source">Source index.</param>
    /// <param name="target">Target index.</param>
    /// <returns>The created or existing link.</returns>
    /// <exception cref="LinkIntegrityError">
    /// Thrown when an explicit index conflicts with an existing structure, is not
    /// a positive integer, or is already in use.
    /// </exception>
    public Link Create(long? index, long source, long target)
    {
        var existing = FindByPair(source, target);
        if (existing is not null)
        {
            if (index is not null && index.Value != existing.Index)
            {
                throw new LinkIntegrityError(
                    string.Create(
                        CultureInfo.InvariantCulture,
                        $"Link ({source} {target}) already exists as {existing.Index}, cannot also be {index.Value}"));
            }

            return existing;
        }

        long id;
        if (index is null)
        {
            id = AllocateIndex();
        }
        else
        {
            id = index.Value;
            if (id < 1)
            {
                throw new LinkIntegrityError("Link index must be a positive integer");
            }

            if (_links.ContainsKey(id))
            {
                throw new LinkIntegrityError(
                    string.Create(CultureInfo.InvariantCulture, $"Link index {id} is already in use"));
            }

            ReserveIndex(id);
        }

        var link = new Link(id, source, target);
        Insert(link);
        return link;
    }

    /// <summary>
    /// Replace the structure of an existing link, preserving its identity unless a
    /// new identity is requested.
    /// </summary>
    /// <param name="index">Identity of the link to update.</param>
    /// <param name="source">New source index.</param>
    /// <param name="target">New target index.</param>
    /// <param name="newIndex">Optional new identity (re-index).</param>
    /// <returns>The updated link.</returns>
    /// <exception cref="LinkIntegrityError">
    /// Thrown when the link is missing, the new structure collides with a
    /// <em>different</em> link, or a requested new index is already in use.
    /// </exception>
    public Link Update(long index, long source, long target, long? newIndex)
    {
        if (!_links.TryGetValue(index, out var current))
        {
            throw new LinkIntegrityError(
                string.Create(CultureInfo.InvariantCulture, $"Cannot update missing link {index}"));
        }

        var id = newIndex ?? index;
        var collision = FindByPair(source, target);
        if (collision is not null && collision.Index != index)
        {
            throw new LinkIntegrityError(
                string.Create(
                    CultureInfo.InvariantCulture,
                    $"Cannot update link {index}: ({source} {target}) already exists as {collision.Index}"));
        }

        if (id != index && _links.ContainsKey(id))
        {
            throw new LinkIntegrityError(
                string.Create(CultureInfo.InvariantCulture, $"Link index {id} is already in use"));
        }

        // Mirror the JavaScript Map semantics exactly: delete the old key then set
        // the new one. In JS, `map.delete(k); map.set(k, v)` re-appends the entry
        // even when the key is unchanged, so `update` always moves the link to the
        // end of iteration order.
        _byPair.Remove((current.Source, current.Target));
        _order.Remove(index);
        _links.Remove(index);

        var link = new Link(id, source, target);
        _links[id] = link;
        _order.Add(id);
        _byPair[(source, target)] = id;
        ReserveIndex(id);
        return link;
    }

    /// <summary>Remove a link by identity.</summary>
    /// <param name="index">Identity to remove.</param>
    /// <returns><see langword="true"/> if a link was removed.</returns>
    public bool Delete(long index)
    {
        if (!_links.TryGetValue(index, out var current))
        {
            return false;
        }

        _byPair.Remove((current.Source, current.Target));
        _order.Remove(index);
        _links.Remove(index);
        return true;
    }

    /// <summary>Remove every link and reset identity allocation.</summary>
    public void Clear()
    {
        _links.Clear();
        _order.Clear();
        _byPair.Clear();
        NextIndex = 1;
    }

    /// <summary>Insert a link, recording insertion order and the pair index.</summary>
    private void Insert(Link link)
    {
        _links[link.Index] = link;
        _order.Add(link.Index);
        _byPair[(link.Source, link.Target)] = link.Index;
    }
}
